### PR TITLE
Remove guzzle dependencies

### PR DESCRIPTION
**Pull request by denniskoster**
**Changes**
- Removed guzzle dependencies
- Made the `MyParcelComApi` and `ClientCredentials` classes guzzle independent. It should now be possible to inject any HTTP client that implements the `php-http/httplug` client.
- Added `php-http/discovery` package to auto discover installed Httpclient

**Notes**
- All this should be tested in a real application (mainly the detection of existing http clients and usage of the adapters) but I'm not sure how to go about doing that

### CODE REVIEW

File: src/Traits/DetectsExistingHttpClient.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        if (function_exists('curl_version')) {
+            // Return Curl client
+            if (!class_exists('Http\Client\Curl\Client')) {
+                throw new MissingHttpClientAdapterException('Missing cURL client package. Run `composer require php-http/curl-client`.');
+            }
+
+            return new Http\Client\Curl\Client();
+        }
+    }
+}
```
Comment (Shodske) [inline src/Traits/DetectsExistingHttpClient.php]: This entire trait is actually unnecessary as you can get `php-http/discovery` to discover a client for you using the static method: `HttpClientDiscovery::find()`

File: src/Authentication/ClientCredentials.php [hunk 2]
```diff
@@ -39,17 +40,27 @@ class ClientCredentials implements AuthenticatorInterface
...
+     * @param ClientInterface|null $httpClient
      */
-    public function __construct($clientId, $clientSecret, $authUri = 'https://sandbox-auth.myparcel.com', CacheInterface $cache = null)
-    {
+    public function __construct(
+        $clientId,
+        $clientSecret,
+        $authUri = 'https://sandbox-auth.myparcel.com',
+        CacheInterface $cache = null,
+        ClientInterface $httpClient = null
```
[the remark this answers is not in the document]
Comment (denniskoster) [reply src/Authentication/ClientCredentials.php]: Damn, I copied the `ClientInterface` from the PSR-18 package, but did not actually want to install that package as it has a hard dependency on php 7. I figured that should work, but of course it's not the same interface the adapters implement, even though their methods are exactly the same.

**Comment by denniskoster** [comment]
- Everything >= 300 should be error
- Use different version of php-http packages, to be able to be compatible with php 5
- Use detection package and remove trait
- Move the httpclient as first parameter, to the second parameter

File: tests/Feature/MyParcelComApiTest.php [hunk 3]
```diff
@@ -52,15 +52,15 @@ public function testSingleton()
...
+        $api = MyParcelComApi::createSingleton($this->authenticator, 'https://sandbox-api.myparcel.com', $this->client);
         $this->assertInstanceOf(MyParcelComApi::class, $api);
         $this->assertEquals($api, MyParcelComApi::getSingleton());
     }
 
     /** @test */
     public function testAuthenticate()
     {
-        $api = new MyParcelComApi();
+        $api = new MyParcelComApi('https://sandbox-api.myparcel.com', $this->client);
```
Comment (M4tini) [inline tests/Feature/MyParcelComApiTest.php]: Why is this test not using `https://api`

File: tests/Feature/MyParcelComApiTest.php [hunk 4]
```diff
@@ -52,15 +52,15 @@ public function testSingleton()
     {
         $this->assertNull(MyParcelComApi::getSingleton());
 
-        $api = MyParcelComApi::createSingleton($this->authenticator);
+        $api = MyParcelComApi::createSingleton($this->authenticator, 'https://sandbox-api.myparcel.com', $this->client);
```
Comment (M4tini) [inline tests/Feature/MyParcelComApiTest.php]: Why is this test not using `https://api`

File: src/Authentication/ClientCredentials.php [hunk 5]
```diff
@@ -103,19 +112,25 @@ protected function authenticate()
...
-        )->then(function (ResponseInterface $response) {
-            $data = \GuzzleHttp\json_decode((string)$response->getBody(), true);
+                self::HEADER_ACCEPT => self::MIME_TYPE_JSONAPI
+            ],
+            $body
+        );
+
+        try {
+            $response = $this->getHttpClient()->sendRequest($request);
+            $data = json_decode((string)$response->getBody(), true);
```
[the remark this answers is not in the document]
Comment (denniskoster) [reply src/Authentication/ClientCredentials.php]: It doesn't appear like it was caught previously, so I'm not sure what to do with it. By default json_decode() will return `null` if it fails

[the remark this answers is not in the document]
Comment (M4tini) [reply src/Authentication/ClientCredentials.php]: Ok, then we leave it like this.

File: tests/Feature/Proxy/CarrierProxyTest.php [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 namespace MyParcelCom\ApiSdk\Tests\Feature\Proxy;
 
-use GuzzleHttp\ClientInterface;
+use MyParcelCom\ApiSdk\Http\Contracts\HttpClient\ClientInterface;
```
Comment (Shodske) [inline tests/Feature/Proxy/CarrierProxyTest.php]: Should this not be `Http\Client\HttpClient`?

It seems like it needs to be changed in a lot of tests.

File: tests/Feature/Proxy/CarrierProxyTest.php [hunk 7]
```diff
@@ -31,7 +31,7 @@ public function setUp()
 
         $this->client = $this->getClientMock();
         $this->authenticator = $this->getAuthenticatorMock();
-        $this->api = (new MyParcelComApi('https://api'))
+        $this->api = (new MyParcelComApi('https://api', $this->client))
             ->setCache(new NullCache())
             ->setHttpClient($this->client)
```
Comment (Shodske) [inline tests/Feature/Proxy/CarrierProxyTest.php]: You're setting the same client twice, once in the constructor and once with the setter.

File: tests/Feature/Proxy/CarrierProxyTest.php (hunk 6 above)
Comment (denniskoster) [reply tests/Feature/Proxy/CarrierProxyTest.php]: That is interesting. I removed the interface with the "safe" method. How are these tests not failing with the interface no longer existing? :confused:

Comment (Shodske) [reply tests/Feature/Proxy/CarrierProxyTest.php]: I think it's because the interface is only used in a docblock.

**Pull request merged**